### DISCUSSION TITLE
Efficiently performing read only operation on entity

### DISCUSSION THREAD

**Discussion by TurakKatase** [Azure/azure-functions-durable-extension#2053] [Q&A] [answered]
I'm using a durable entity to determine the enabled status of various items. For example, at a couple points in an orchestration, I check to make sure that the ID of the task isn't disabled. If it is disabled, I send the orchestration into a standby loop until it is enabled again.

Right now, I have an `IsEnabled` method on my entity, along with the various methods to update the entity status. All of this works well. But my calls to `IsEnabled` appear to be a bottleneck. It's called fairly frequently from numerous parallel orchestrations.

The documentation states 
> To prevent conflicts, all operations on a single entity are guaranteed to execute serially, that is, one after another.

As this is a read-only check, this seems inefficient. I can't use the the `IDurableEntityClient` inside the orchestration to get the read-only state. Is there any more efficient way of doing this that would allow me to perform the checks in parallel?

**Comment by cgillum** [comment, marked as answer]
I think the reason this creates a performance bottleneck is because each `IsEnabled` check from your orchestration(s) to the entity is represented as a series of messages that need to be processed by the entity. The entity can only process one message at a time, which means it slows down as you add more orchestrations.

You might be able to speed this up quite a bit by using `IDurableEntityClient` in an activity function that fetches the read-only state of the entity (which is very fast). It still involves messaging because you have to invoke an activity function, but it's more scalable since it removes the entity operation throughput bottleneck. This option should be viable as long as you can tolerate potentially stale information.

Adding @sebastianburckhardt in case he has additional thoughts.

**Reply by TurakKatase** [reply]
Thanks for the suggestion. That makes sense and seems to have helped a bit on a relatively small number of parallel orchestrations. It should help more when I can run a test with a larger number of parallel orchestrations.
